api/src/main/java/org/pmiops/workbench/exfiltration/EgressEventServiceImpl.java

```diff
@@ -73,22 +84,62 @@ public void handleEvent(SumologicEgressEvent event) {
+    }
+
+    for (Optional<DbUser> egressUserMaybe : egressUsers) {
+      logger.warning(
+          String.format(
+              "Received an egress event from workspace namespace %s, googleProject %s (%.2fMiB, VM name %s)",
+              workspaceNamespace, event.getProjectName(), event.getEgressMib(), event.getVmName()));
+      this.egressEventAuditor.fireEgressEventForUser(event, egressUserMaybe.orElse(null));
+
+      Optional<DbEgressEvent> maybeEvent =
+          this.maybePersistEgressEvent(event, egressUserMaybe, dbWorkspaceMaybe);
+      maybeEvent.ifPresent(e -> taskQueueService.pushEgressEventTask(e.getEgressEventId()));
+    }
+  }
+
+  @NotNull
+  private List<Optional<DbUser>> findEgressUsers(
+      SumologicEgressEvent event, Optional<DbWorkspace> dbWorkspace) {
+
+    // This case is when the Egress is from a GCE cluster.
+    if (StringUtils.isNotEmpty(event.getVmPrefix())
+        && event.getVmPrefix().startsWith("all-of-us")) {
+      Optional<DbUser> dbUserMaybe =
+          vmNameToUserDatabaseId(event.getVmPrefix()).flatMap(userService::getByDatabaseId);
+      if (!dbUserMaybe.isPresent()) {
+        logger.warning(String.format("User not found by given VM prefix: %s", event.getVmPrefix()));
+      }
+      return Collections.singletonList(dbUserMaybe);
+    }
 
-    Optional<DbUser> dbUserMaybe =
-        vmNameToUserDatabaseId(event.getVmPrefix()).flatMap(userService::getByDatabaseId);
-    if (!dbUserMaybe.isPresent()) {
-      logger.warning(String.format("user not found by given VM prefix: %s", event.getVmPrefix()));
+    // This is the GKE case
+    if (dbWorkspace.isPresent() && StringUtils.isNotEmpty(event.getSrcGkeCluster())) {
+      List<Optional<DbUser>> appUsers = findAppUsers(dbWorkspace.get());
+      return appUsers;
     }
 
-    logger.warning(
-        String.format(
-            "Received an egress event from workspace namespace %s, googleProject %s (%.2fMiB, VM prefix %s)",
-            workspaceNamespace, event.getProjectName(), event.getEgressMib(), event.getVmPrefix()));
-    this.egressEventAuditor.fireEgressEvent(event);
+    return Collections.emptyList();
+  }
 
-    Optional<DbEgressEvent> maybeEvent =
-        this.maybePersistEgressEvent(event, dbUserMaybe, dbWorkspaceMaybe);
-    maybeEvent.ifPresent(e -> taskQueueService.pushEgressEventTask(e.getEgressEventId()));
+  private List<Optional<DbUser>> findAppUsers(DbWorkspace dbWorkspace) {
+    List<UserAppEnvironment> userAppEnvironments =
+        leonardoApiClient.listAppsInProjectAsService(dbWorkspace.getGoogleProject());
+    List<String> appUsersList =
+        userAppEnvironments.stream()
+            .filter(u -> u.getStatus().equals(AppStatus.RUNNING))
+            .map(u -> u.getCreator())
+            .collect(Collectors.toList());
+    List<DbUser> usersByUsernames = userService.findUsersByUsernames(appUsersList);
+    return usersByUsernames.stream().map(Optional::of).collect(Collectors.toList());
```

Comment:
Is using list of optional redundant/anti-pattern?
Seems findUsersByUsernames will not return null. Even so, usually we can just use list that contains non-null.